File: packages/quran-engine-py/quran_engine/alphabet.py
"""The Arabic alphabet as a Quran reader meets it: every letter with its joining forms, its name
and transliteration, and - the part that matters for tajweed - its WEIGHT.

Weight is why this belongs in a tajweed engine rather than in a phrasebook. Every letter is
pronounced thin (tarqiq) or full (tafkhim), a few depend on context (raa, and the lam of the divine
name), and alif has no weight of its own at all: it inherits the letter before it. That single fact
is behind a large share of beginner mistakes, and it is a property of the letter, not of any
particular verse, so it lives here beside the letter and not in the annotation corpus.

Also carried: the letters outside the 28 (hamza, ta marbuta, lam-alif), the six Persian/Urdu letters
that appear in some printed mushafs, the Eastern-Arabic numerals the ayah markers use, the tashkeel
marks, and the waqf (stopping) signs.

See ../../docs/16-arabic-alphabet.md.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional


@dataclass(frozen=True)
class ArabicLetter:
    id: int
    letter: str
    forms: list[str]
    name: str
    transliteration: str
    show_tashkeel: bool
    sound: str
    #: "light" | "heavy" | "conditional" | "followsPrevious", or None where none is recorded.
    weight: Optional[str] = None
    #: Why, in one sentence.
    weight_rule: Optional[str] = None

    @staticmethod
    def from_json(raw: dict) -> "ArabicLetter":
        return ArabicLetter(
            id=raw.get("id", 0),
            letter=raw.get("letter", ""),
            forms=list(raw.get("forms") or []),
            name=raw.get("name", ""),
            transliteration=raw.get("transliteration", ""),
            show_tashkeel=bool(raw.get("showTashkeel", False)),
            sound=raw.get("sound", ""),
            weight=raw.get("weight"),
            weight_rule=raw.get("weightRule"),
        )


@dataclass(frozen=True)
class Tashkeel:
    english: str
    arabic: str
    mark: str
    transliteration: str


@dataclass(frozen=True)
class StoppingSign:
    symbol: str
    title: str


@dataclass(frozen=True)
class ArabicNumeral:
    number: str
    name: str
    transliteration: str
    english_number: str


class ArabicAlphabet:
    def __init__(self, data: Optional[dict] = None) -> None:
        self._data = data or {}

    def letters(self) -> list[ArabicLetter]:
        """The 28 letters of the alphabet, in order."""
        return [ArabicLetter.from_json(r) for r in self._data.get("standardLetters", [])]

    def other_letters(self) -> list[ArabicLetter]:
        """Hamza, ta marbuta, lam-alif and the rest: written forms outside the 28."""
        return [ArabicLetter.from_json(r) for r in self._data.get("otherLetters", [])]

    def non_arabic_script_letters(self) -> list[ArabicLetter]:
        """The Persian/Urdu letters some printed mushafs use for non-Arabic sounds."""
        return [ArabicLetter.from_json(r) for r in self._data.get("nonArabicScriptLetters", [])]

    def all_letters(self) -> list[ArabicLetter]:
        """Every letter this reference knows, the 28 first."""
        return self.letters() + self.other_letters() + self.non_arabic_script_letters()

    def letter(self, letter: str) -> Optional[ArabicLetter]:
        """One letter by its isolated form. Accepts any of its joining forms too, so a letter
        lifted out of a word still resolves."""
        wanted = letter.strip()
        if not wanted:
            return None
        for entry in self.all_letters():
            if entry.letter == wanted or wanted in entry.forms:
                return entry
        return None

    def letter_by_id(self, letter_id: int) -> Optional[ArabicLetter]:
        for entry in self.all_letters():
            if entry.id == letter_id:
                return entry
        return None

    def weight(self, letter: str) -> Optional[str]:
        """The tajweed weight of a letter, or None when the reference records none."""
        entry = self.letter(letter)
        return entry.weight if entry else None

    def weight_descriptions(self) -> dict[str, str]:
        """What a weight name means, in one line."""
        return dict(self._data.get("weights") or {})

    def heavy_letters(self) -> list[ArabicLetter]:
        """The letters pronounced full - the isti'la letters."""
        return [entry for entry in self.letters() if entry.weight == "heavy"]

    def tashkeel(self) -> list[Tashkeel]:
        return [Tashkeel(r.get("english", ""), r.get("arabic", ""), r.get("mark", ""),
                         r.get("transliteration", ""))
                for r in self._data.get("tashkeel", [])]

    def stopping_signs(self) -> list[StoppingSign]:
        """The waqf signs, with what each one tells the reciter to do."""
        return [StoppingSign(r.get("symbol", ""), r.get("title", ""))
                for r in self._data.get("stoppingSigns", [])]

    def stopping_sign(self, symbol: str) -> Optional[StoppingSign]:
        for sign in self.stopping_signs():
            if sign.symbol == symbol:
                return sign
        return None

    def numbers(self) -> list[ArabicNumeral]:
        """The Eastern-Arabic numerals, 0 through 10."""
        return [ArabicNumeral(r.get("number", ""), r.get("name", ""),
                              r.get("transliteration", ""), r.get("englishNumber", ""))
                for r in self._data.get("numbers", [])]

    def stopping_signs_source(self) -> str:
        return self._data.get("stoppingSignsSource", "")

File: packages/quran-engine-rust/src/alphabet.rs
//! The Arabic alphabet as a Quran reader meets it: every letter with its joining forms, its name and
//! transliteration, and — the part that matters for tajweed — its WEIGHT.
//!
//! Weight is why this belongs in a tajweed engine rather than in a phrasebook. Every letter is
//! pronounced thin (tarqiq) or full (tafkhim), a few depend on context (raa, and the lam of the
//! divine name), and alif has no weight of its own at all: it inherits the letter before it. That
//! single fact is behind a large share of beginner mistakes, and it is a property of the letter, not
//! of any particular verse, so it lives here beside the letter and not in the annotation corpus.
//!
//! Also carried: the letters outside the 28, the six Persian/Urdu letters some printed mushafs use,
//! the Eastern-Arabic numerals, the tashkeel marks, and the waqf (stopping) signs.
//!
//! See `../../docs/16-arabic-alphabet.md`.

use serde::Deserialize;
use std::collections::HashMap;

#[derive(Debug, Clone, Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "camelCase")]
pub struct ArabicLetter {
    #[serde(default)]
    pub id: u32,
    #[serde(default)]
    pub letter: String,
    #[serde(default)]
    pub forms: Vec<String>,
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub transliteration: String,
    #[serde(default)]
    pub show_tashkeel: bool,
    #[serde(default)]
    pub sound: String,
    /// `"light"`, `"heavy"`, `"conditional"`, `"followsPrevious"` — or `None` where none is recorded.
    #[serde(default)]
    pub weight: Option<String>,
    /// Why, in one sentence.
    #[serde(default)]
    pub weight_rule: Option<String>,
}

#[derive(Debug, Clone, Deserialize, PartialEq, Eq, Default)]
pub struct Tashkeel {
    #[serde(default)]
    pub english: String,
    #[serde(default)]
    pub arabic: String,
    #[serde(default)]
    pub mark: String,
    #[serde(default)]
    pub transliteration: String,
}

#[derive(Debug, Clone, Deserialize, PartialEq, Eq, Default)]
pub struct StoppingSign {
    #[serde(default)]
    pub symbol: String,
    #[serde(default)]
    pub title: String,
}

#[derive(Debug, Clone, Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "camelCase")]
pub struct ArabicNumeral {
    #[serde(default)]
    pub number: String,
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub transliteration: String,
    #[serde(default)]
    pub english_number: String,
}

/// `data/arabic-alphabet.json`.
#[derive(Debug, Clone, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ArabicAlphabetFile {
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub weights: HashMap<String, String>,
    #[serde(default)]
    pub standard_letters: Vec<ArabicLetter>,
    #[serde(default)]
    pub other_letters: Vec<ArabicLetter>,
    #[serde(default)]
    pub non_arabic_script_letters: Vec<ArabicLetter>,
    #[serde(default)]
    pub numbers: Vec<ArabicNumeral>,
    #[serde(default)]
    pub tashkeel: Vec<Tashkeel>,
    #[serde(default)]
    pub stopping_signs: Vec<StoppingSign>,
    #[serde(default)]
    pub stopping_signs_source: String,
}

use crate::Engine;

impl Engine {
    /// The 28 letters of the alphabet, in order.
    pub fn letters(&self) -> &[ArabicLetter] {
        &self.alphabet.standard_letters
    }

    /// Hamza, ta marbuta, lam-alif and the rest: written forms outside the 28.
    pub fn other_letters(&self) -> &[ArabicLetter] {
        &self.alphabet.other_letters
    }

    /// The Persian/Urdu letters some printed mushafs use for non-Arabic sounds.
    pub fn non_arabic_script_letters(&self) -> &[ArabicLetter] {
        &self.alphabet.non_arabic_script_letters
    }

    /// Every letter this reference knows, the 28 first.
    pub fn all_letters(&self) -> Vec<&ArabicLetter> {
        self.letters()
            .iter()
            .chain(self.other_letters())
            .chain(self.non_arabic_script_letters())
            .collect()
    }

    /// One letter by its isolated form. Accepts any of its joining forms too, so a letter lifted out
    /// of a word still resolves.
    pub fn letter(&self, letter: &str) -> Option<&ArabicLetter> {
        let wanted = letter.trim();
        if wanted.is_empty() {
            return None;
        }
        self.all_letters()
            .into_iter()
            .find(|entry| entry.letter == wanted || entry.forms.iter().any(|f| f == wanted))
    }

    pub fn letter_by_id(&self, id: u32) -> Option<&ArabicLetter> {
        self.all_letters().into_iter().find(|entry| entry.id == id)
    }

    /// The tajweed weight of a letter, or `None` when the reference records none.
    pub fn letter_weight(&self, letter: &str) -> Option<&str> {
        self.letter(letter)?.weight.as_deref()
    }

    /// What a weight name means, in one line.
    pub fn weight_descriptions(&self) -> &HashMap<String, String> {
        &self.alphabet.weights
    }

    /// The letters pronounced full — the isti'la letters.
    pub fn heavy_letters(&self) -> Vec<&ArabicLetter> {
        self.letters().iter().filter(|l| l.weight.as_deref() == Some("heavy")).collect()
    }

    pub fn tashkeel(&self) -> &[Tashkeel] {
        &self.alphabet.tashkeel
    }

    /// The waqf signs, with what each one tells the reciter to do.
    pub fn stopping_signs(&self) -> &[StoppingSign] {
        &self.alphabet.stopping_signs
    }

    pub fn stopping_sign(&self, symbol: &str) -> Option<&StoppingSign> {
        self.stopping_signs().iter().find(|sign| sign.symbol == symbol)
    }

    /// The Eastern-Arabic numerals, 0 through 10.
    pub fn arabic_numbers(&self) -> &[ArabicNumeral] {
        &self.alphabet.numbers
    }

    pub fn stopping_signs_source(&self) -> &str {
        &self.alphabet.stopping_signs_source
    }
}
